### PR TITLE
INT-1503 Changes in Affirm Strategy

### PR DESCRIPTION
**Pull request by BC-CesarGutierrez**
[INT 1503](https://jira.bigcommerce.com/browse/INT-1503)
## What?

- Include category names in each product added to an order.

## Why?
After Affirm's team review in test store, this changes were required to complete validation on the payment provider integration

## Sibling PR
https://github.com/bigcommerce/bigcommerce/pull/29921

## Testing / Proof
https://drive.google.com/file/d/18wCExeuKXw_QukQgVlA-idy5P3Nt1N6C/view?usp=sharing

@bigcommerce/checkout @bigcommerce/payments

### CODE REVIEW

File: src/payment/strategies/affirm/affirm-payment-strategy.ts [hunk 1]
```diff
@@ -126,17 +128,19 @@ export default class AffirmPaymentStrategy implements PaymentStrategy {
...
+                platform_type: 'bigcommerce',
+                platform_version: '',
+                platform_affirm: '',
             },
+            discounts: this._getDiscounts(),
             order_id: order.orderId ? order.orderId.toString() : '',
-            shipping_amount: order.shippingCostTotal * 100,
-            tax_amount: order.taxTotal * 100,
-            total: order.orderAmount * 100,
+            shipping_amount: round(order.shippingCostTotal * 100, 0),
```
Comment (davidchin) [inline src/payment/strategies/affirm/affirm-payment-strategy.ts]: Ideally, we shouldn't be converting dollars to cents like this. This is because the conversion is not always accurate as you're dealing with floats. Also, not all currencies have their main unit (i.e.: dollar) divided into 100 subunits (cents). Instead, we should ask the backend to return the original / integer amount directly. In here, we can simply forward the amount to Affirm. `orderAmountAsInteger` is an example of that approach.

File: src/payment/strategies/affirm/affirm-payment-strategy.ts [hunk 2]
```diff
@@ -126,17 +128,19 @@ export default class AffirmPaymentStrategy implements PaymentStrategy {
             shipping: this._getShippingAddress(),
             billing: this._getBillingAddress(),
             items: this._getItems(),
-            discounts: this._getDiscounts(),
             metadata: {
                 shipping_type: consignment.selectedShippingOption.type,
                 mode: 'modal',
+                platform_type: 'bigcommerce',
+                platform_version: '',
```
[the remark this answers is not in the document]
Comment (michxff) [reply src/payment/strategies/affirm/affirm-payment-strategy.ts]: Affirm requested the following changes

>  In our case,

      "platform_type": "BigCommerce" (mandatory)
      "platform_version": refers to the version of BigCommerce the customer is on (if applicable)
      "platform_affirm": the version of the BigCommerce/Affirm integration. If there is no versioning in place, leave blank.

Do we have the bigcommerce version available at the checkout-sdk ? otherwise I think we can leave it blank.

**Comment by michxff** [comment]
Hi @davidchin thank you for your feedback, @BC-CesarGutierrez did some changes on this PR to receive the amounts in cents from the backend directly

A PR in bcapp was created to return the values in the expected format (https://github.com/bigcommerce/bigcommerce/pull/29857)

can you please take a look ♻️?

File: src/discount/discounts.mock.ts [hunk 3]
```diff
@@ -4,5 +4,6 @@ export function getDiscount(): Discount {
     return {
         id: '12e11c8f-7dce-4da3-9413-b649533f8bad',
         discountedAmount: 10,
+        discountedAmountAsInteger: 10,
```
Comment (michxff) [inline src/discount/discounts.mock.ts]: can we add correct values for the Integer values 10 => 1000, this applies to all the mocked values which represent integers

File: src/payment/strategies/affirm/affirm-payment-strategy.ts [hunk 4]
```diff
@@ -126,17 +126,17 @@ export default class AffirmPaymentStrategy implements PaymentStrategy {
             shipping: this._getShippingAddress(),
             billing: this._getBillingAddress(),
             items: this._getItems(),
-            discounts: this._getDiscounts(),
             metadata: {
                 shipping_type: consignment.selectedShippingOption.type,
                 mode: 'modal',
+                platform_type: 'bigcommerce',
```
Comment (michxff) [inline src/payment/strategies/affirm/affirm-payment-strategy.ts]: platform_type value must be BigCommerce, the other two values related to the platform must be kept in blank

**Comment by BC-CesarGutierrez** [comment]
Comments have been addressed @michxff ♻️

**Comment by michxff** [comment]
@BC-CesarGutierrez could you please add this pr https://github.com/bigcommerce/bigcommerce/pull/29857 as sibling, and add the video proof?

**Comment by BC-CesarGutierrez** [comment]
PR requested changes done @michxff ♻️

**Comment by JoseBobadilla20** [comment]
💚 Tested

File: src/payment/strategies/affirm/affirm-payment-strategy.ts [hunk 5]
```diff
@@ -126,17 +126,19 @@ export default class AffirmPaymentStrategy implements PaymentStrategy {
...
+                platform_type: 'BigCommerce',
+                platform_version: '',
+                platform_affirm: '',
             },
+            discounts: this._getDiscounts(),
             order_id: order.orderId ? order.orderId.toString() : '',
-            shipping_amount: order.shippingCostTotal * 100,
-            tax_amount: order.taxTotal * 100,
-            total: order.orderAmount * 100,
+            shipping_amount: order.shippingCostTotalAsInteger,
```
Comment (davidchin) [inline src/payment/strategies/affirm/affirm-payment-strategy.ts]: You'll need to use `AmountTransformer` to do the dollar to cent conversion. It takes into consideration of the decimal places of the transacting currency.

File: src/cart/internal-line-item.ts [hunk 6]
```diff
@@ -12,7 +12,7 @@ export default interface InternalLineItem {
     name?: string;
     quantity: number;
     brand?: string;
-    categoryNames?: string[];
+    categoryNames?: Array<[string]>;
```
Comment (davidchin) [inline src/cart/internal-line-item.ts]: `string[][]`

File: src/payment/strategies/affirm/affirm-payment-strategy.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidchin) [reply src/payment/strategies/affirm/affirm-payment-strategy.ts]: I don't think we have. When you say you can leave it blank, do you mean sending an empty string or leaving it out?

[the remark this answers is not in the document]
Comment (michxff) [reply src/payment/strategies/affirm/affirm-payment-strategy.ts]: @davidchin We can remove it from the payload meanwhile. However, I sent an email to the Affirm folks asking about this to confirm.

File: src/order/map-to-internal-order.ts [hunk 7]
```diff
@@ -69,7 +69,7 @@ export default function mapToInternalOrder(order: Order, orderMeta: OrderMetaSta
         },
         grandTotal: {
             amount: order.orderAmount,
-            integerAmount: order.orderAmountAsInteger,
+            integerAmount: amountTransformer.toInteger(order.orderAmount),
```
Comment (michxff) [inline src/order/map-to-internal-order.ts]: is this change needed? we are not doing any change on the backend fields

File: src/order/order.ts [hunk 8]
```diff
@@ -20,7 +20,6 @@ export default interface Order {
     isTaxIncluded: boolean;
     lineItems: LineItemMap;
     orderAmount: number;
-    orderAmountAsInteger: number;
```
Comment (michxff) [inline src/order/order.ts]: same here

File: src/order/orders.mock.ts [hunk 9]
```diff
@@ -48,7 +48,6 @@ export function getOrder(): Order {
         shippingCostBeforeDiscount: 20,
         handlingCostTotal: 8,
         orderAmount: 190,
-        orderAmountAsInteger: 19000,
```
Comment (michxff) [inline src/order/orders.mock.ts]: same here

File: src/payment/strategies/affirm/affirm-payment-strategy.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (michxff) [reply src/payment/strategies/affirm/affirm-payment-strategy.ts]: @davidchin Affirm confirmed that we must sent the platform_version & platform_affirm as **empty strings**

File: src/order/map-to-internal-order.ts [hunk 10]
```diff
@@ -69,7 +69,7 @@ export default function mapToInternalOrder(order: Order, orderMeta: OrderMetaSta
         },
         grandTotal: {
             amount: order.orderAmount,
-            integerAmount: order.orderAmountAsInteger,
+            integerAmount: order.orderAmount,
```
Comment (michxff) [inline src/order/map-to-internal-order.ts]: please remove this change as well

**Comment by BC-CesarGutierrez** [comment]
Changes addressed @michxff ♻️

File: src/payment/strategies/affirm/affirm-payment-strategy.ts [hunk 11]
```diff
@@ -256,27 +253,44 @@ export default class AffirmPaymentStrategy implements PaymentStrategy {
...
+            discounts.discounts = {
+                discount_amount: transformer.toInteger(order.discountAmount),
+                discount_display_name: 'discount',
             };
         }
 
         return discounts;
     }
 
+    private _getTransformer(): AmountTransformer {
```
Comment (michxff) [inline src/payment/strategies/affirm/affirm-payment-strategy.ts]: I think the AmountTransformer should be initialized at the constructor, or received as a constructor parameter, what do you think @davidchin ?

**Comment by michxff** [comment]
@BC-CesarGutierrez @davidchin This PR was updated to include only the category functionality, the change related to the amounts and the metadata information is now on https://github.com/bigcommerce/checkout-sdk-js/pull/568

File: src/cart/internal-carts.mock.ts [hunk 12]
```diff
@@ -27,7 +27,7 @@ export function getCart(): InternalCart {
                 variantId: 71,
                 addedByPromotion: false,
                 productId: 103,
-                categoryNames: ['Cat 1'],
+                categoryNames: [['Cat 1']],
```
Comment (Sure-Round) [inline src/cart/internal-carts.mock.ts]: Would probably be good for one of our tests to have more than 1 category name

**Comment by BC-CesarGutierrez** [comment]
@Sure-Round tests have been updated  ♻️

**Comment by michxff** [comment]
Hi @davidchin , just to confirm 

1. We are going to add the optional include (I'm assuming that those are the ones passed as `include=` parameters on the storefront requests), right?

2. I didn't get completely what we need to do to accomplish the last part, what is the meaning of OOPC?, as I understood we need to add the categories include only when Affirm is the payment method used to create the order, can you please confirm if it is correct or provide some guidance about this?

If you can provide a high-level list of task needed to accomplish this task will be very helpful

**Comment by davidchin** [comment]
1. Yes, that’s correct.
2. Sorry about the acronym. OOPC (Optimised One Page Checkout) is just another name for UCO. And yes, regarding the last part - you need to send that `include` parameter, which is currently missing in the PR at the moment.

**Comment by BC-CesarGutierrez** [comment]
I have included the missing references for categoryNames @davidchin ♻️

File: src/cart/line-items.mock.ts [hunk 13]
```diff
@@ -29,6 +29,7 @@ export function getPhysicalItem(): PhysicalItem {
                 valueId: 3,
             },
         ],
+        categories: [[{name: 'Cat 1'}], [{name: 'Cat 2'}], [{name: 'Cat 3'}]],
```
Comment (davidchin) [inline src/cart/line-items.mock.ts]: I think it might be nice if you can provide multiple category names for a category. i.e.: `[{ name: 'Furniture' }, { name: 'Bed' }]`. That way, you can better test your `map` function.

File: src/checkout/checkout-params.ts [hunk 14]
```diff
@@ -1,7 +1,14 @@
 export enum CheckoutIncludes {
     AvailableShippingOptions = 'consignments.availableShippingOptions',
-    PhysicalItemsCategoryNames = 'cart.lineItems.physicalItems.categoryNames',
+    DigitalItemsCategories = 'cart.lineItems.digitalItems.categories',
     DigitalItemsCategoryNames = 'cart.lineItems.digitalItems.categoryNames',
+    DigitalItemsSocialMedia =  'lineItems.digitalItems.socialMedia',
```
Comment (michxff) [inline src/checkout/checkout-params.ts]: You are mixing checkout includes with order includes

File: src/common/http-request/request-options.ts [hunk 15]
```diff
@@ -15,4 +15,5 @@ export default interface RequestOptions<TParams = {}> {
      * The parameters of the request, if required.
      */
     params?: TParams;
+
```
Comment (michxff) [inline src/common/http-request/request-options.ts]: can you please remove this line

File: src/order/order-request-sender.spec.ts [hunk 16]
```diff
@@ -1,24 +1,24 @@
-import { createRequestSender, createTimeout, Response } from '@bigcommerce/request-sender';
+import {createRequestSender, createTimeout, Response} from '@bigcommerce/request-sender';
```
Comment (michxff) [inline src/order/order-request-sender.spec.ts]: why spaces where removed from the includes?

File: src/order/order-request-sender.ts [hunk 17]
```diff
@@ -1,31 +1,30 @@
-import { RequestSender, Response } from '@bigcommerce/request-sender';
+import {RequestSender, Response} from '@bigcommerce/request-sender';
```
Comment (michxff) [inline src/order/order-request-sender.ts]: can you please re-add spaces?

File: src/order/order-request-sender.ts [hunk 18]
```diff
@@ -1,31 +1,30 @@
...
+            CheckoutIncludes.Payments,
+            CheckoutIncludes.PhysicalItemsSocialMedia,
+            CheckoutIncludes.PhysicalItemsOptions,
+            CheckoutIncludes.DigitalItemsSocialMedia,
+            CheckoutIncludes.DigitalItemsOptions,
+        ];
 
         return this._requestSender.get(url, {
-            params,
+            params : {include : params && params.include ? include.concat(params.include).join(',') : include.join(',')},
```
Comment (michxff) [inline src/order/order-request-sender.ts]: please follow the `checkout-request-sender.ts` approach

File: src/payment/strategies/affirm/affirm-payment-strategy.ts [hunk 19]
```diff
@@ -57,7 +59,19 @@ export default class AffirmPaymentStrategy implements PaymentStrategy {
             throw new PaymentArgumentInvalidError(['payment.methodId']);
         }
 
-        return this._store.dispatch(this._orderActionCreator.submitOrder({ useStoreCredit }, options))
+        const requestOptions: PaymentRequestOptions = {
```
Comment (michxff) [inline src/payment/strategies/affirm/affirm-payment-strategy.ts]: this could be refactored, I will check how

Comment (michxff) [reply src/payment/strategies/affirm/affirm-payment-strategy.ts]: I think you can do something like

```typescript
{
    ...options,
    params: {
        include: [
            'cart.lineItems.physicalItems.categories',
            'cart.lineItems.digitalItems.categories',
        ],
    }
}
```

File: src/checkout/checkout-params.ts [hunk 20]
```diff
@@ -1,7 +1,7 @@
 export enum CheckoutIncludes {
     AvailableShippingOptions = 'consignments.availableShippingOptions',
-    PhysicalItemsCategoryNames = 'cart.lineItems.physicalItems.categoryNames',
     DigitalItemsCategoryNames = 'cart.lineItems.digitalItems.categoryNames',
+    PhysicalItemsCategoryNames = 'cart.lineItems.physicalItems.categoryNames',
```
Comment (michxff) [inline src/checkout/checkout-params.ts]: can you please revert this change?

**Comment by michxff** [comment]
thanks @davidchin for your feedback, this PR is ready for another look ♻️

File: src/checkout/checkout-params.ts [hunk 21]
```diff
@@ -2,6 +2,9 @@ export enum CheckoutIncludes {
     AvailableShippingOptions = 'consignments.availableShippingOptions',
     PhysicalItemsCategoryNames = 'cart.lineItems.physicalItems.categoryNames',
     DigitalItemsCategoryNames = 'cart.lineItems.digitalItems.categoryNames',
+    OrderPayments = 'payments',
```
Comment (davidchin) [inline src/checkout/checkout-params.ts]: You'll need to define a new `OrderParams` interface. `CheckoutParams` is only for checkout endpoints.

Comment (davidchin) [reply src/checkout/checkout-params.ts]: Also, you don't need to add `payments` to `OrderParams `.

File: src/order/order-request-sender.ts [hunk 22]
```diff
@@ -11,21 +12,23 @@ export default class OrderRequestSender {
...
+            'lineItems.digitalItems.options',
+        ];
 
         return this._requestSender.get(url, {
-            params,
+            params: {
+                include: joinIncludes([
+                    ...include,
+                    ...(params && params.include || []),
+                ])},
```
Comment (davidchin) [inline src/order/order-request-sender.ts]: `}` needs to be on a new line 💅

File: src/payment/strategies/affirm/affirm-payment-strategy.ts [hunk 23]
```diff
@@ -57,7 +58,17 @@ export default class AffirmPaymentStrategy implements PaymentStrategy {
             throw new PaymentArgumentInvalidError(['payment.methodId']);
         }
 
-        return this._store.dispatch(this._orderActionCreator.submitOrder({ useStoreCredit }, options))
+        const requestOptions = {
+            ...options,
+            params: {
+                include: [
+                    'lineItems.physicalItems.categories',
```
Comment (davidchin) [inline src/payment/strategies/affirm/affirm-payment-strategy.ts]: You can use `OrderIncludes.DigitalItemsCategories` and `OrderIncludes.PhysicalItemsCategories`

File: src/payment/strategies/affirm/affirm-payment-strategy.ts [hunk 24]
```diff
@@ -277,4 +291,16 @@ export default class AffirmPaymentStrategy implements PaymentStrategy {
...
         return discounts;
     }
+
+    private _getCategories(categories?: LineItemCategory[][]): string[][] {
+        if (!categories) {
+            return[[]];
+        }
+
+        return categories.map(categoryTree => {
+            return categoryTree.map( category => {
```
Comment (michxff) [inline src/payment/strategies/affirm/affirm-payment-strategy.ts]: There is an extra space before the category variable, you can also use the arrow function short syntax

```
categories.map(categoryTree => categoryTree.map(category => category.name))
```

Comment (BC-CesarGutierrez) [reply src/payment/strategies/affirm/affirm-payment-strategy.ts]: 👍🏼

**Comment by michxff** [comment]
Thanks @BC-CesarGutierrez can you please add the sibling section in the description of this PR?

**Comment by BC-CesarGutierrez** [comment]
Sibling section added @michxff

**Comment by BC-CesarGutierrez** [comment]
Comments have been addressed @davidchin  ♻️

**Comment by davidbarocio** [comment]
💚 Tested
